### ISSUE TITLE
Finetune models with different images sizes and partially labeled multi-class masks

### ISSUE THREAD

**Issue by WeisiX** [microsoft/BiomedParse#62]
Hello @theodore-zhao and other authors, thank you again for this amazing and well-documented work! I was able to run the inference and tried some finetuning on my own RGB pathology images. To improve my usage, I have two questions:

1) Is it possible to fine-tune the existing model with images that are not 1024 by 1024 pixels? I have images that are about 1024 width and 512 height, so padding the height will results in a lot of unnecessary background 0 intensity pixels in my training dataset, and I don't want to reshape my images to 1024 by 1024 either because keep the resolution consistent is important for my case. Why is there a current requirement of 1024 by 1024 image size, and if training with other image sizes are allowed, what do I need to change in the code or script to make that happen?

2) How does the training take into account the background pixels in the training data? My training data was partially labeled, so not all pixels have foreground labels, but there are pixels in the background that could actually belong to one of the classes. Is there a way to make the model not learn from the background and only the foreground in the multi-class masks?

Thank you!

**Comment by theodore-zhao** [comment]
1. It is possible to run finetuning with a different resolution, however, it is recommended that the height and width are the same, e.g. 512x512 or 1024x1024. This is because the aseptic ratio was kept 1:1 for all stages of training which was also a default in some parts of the model architecture. Padding was used when preprocessing the BiomedParse datasets, so it won't cause performance degradation, but as you said, is less efficient. Another way for your case maybe cutting the image into 2 square images of 512x512, if that setting makes sense for your application.
2. The model is a text prompted segmentation, and the multi-class setting is handled by the text prompt. Given a prompt, the model always does binary segmentation on the pixels belonging to the described class. The data annotation doesn't need to cover all pixels in that case. All you need to do is to specify the class in text prompt, and use the mask for this class as the binary segmentation target. Different text prompts can be used on the same image to perform segmentation for different classes.

**Comment by WeisiX** [comment]
Thank you for the response! I took your advice and created a 512x512 training pathology dataset, and had multiple binary masks representing multi-class labels for each image. The evaluation after last epoch of fine tuning gave a 62.5 mIoU score. I then tested on new images. An interesting observation is that when I gave it text prompt to objects that are not present in the image (e.g., another class that I added in training set but nonexistent in that particular image), the model still predicts a mask for it, and usually this mask is almost identical to one of the other classes that do exist in that image. 

What could be the cause of this? Is it related to the fact that the model treats each class as a binary segmentation and therefore could have some bias?

I also read from https://github.com/microsoft/BiomedParse/issues/21 that we could use `check_mask_stats` to filter out unlikely masks, but to my understanding it would only work if the target is not likely to be present in the *modality* (which is my case is not true). Also for self-defined new classes/target, it doesn't seems to support the usage.

Any suggestions for the false-positives in this case? Would appreciate your insights!

**Comment by theodore-zhao** [comment]
Glad to hear that the finetuning was working! 

Yes, your observation is one common behavior of the model. By design the model was trained to "segment the most probable region corresponding to the text prompt". Therefore, the model will tend to segment something that is most likely based on its training data. The way we took to solve this was exactly using check_mask_stats to filter out unlikely masks. Although the initial purpose was to filter out invalid segmentation that are very different such as a different modality, it works for same modality as well.

Also we don't have supports for your specific use case, here is a brief summary of how you can do it for your dataset.
1. Run model inference on the dataset that the model was finetuned on, and get the predicted mask for a specific object type.
2. For each image, calculate the stats by averaging the following four values on the predicted region: pixelwise probability (scaled to 0-255), and the R, G, B values of the original image.
4. Fit a beta distribution for each of the 4 values over the dataset, and that is the unique parameter for this object type.
5. When the model is segmenting some mask, calculate the stats and compare with the beta distribution to determine if it belongs to this object type.

This is a train-free way that doesn't involve any training with negative examples. Hope it works for you!

**Comment by WeisiX** [comment]
Thank you @theodore-zhao, as a feedback to your suggestions: I was able to generate my own beta distribution parameters for new classes in pathology images, but it did not work too well to reduce the false positives. I think this is probably due to the image-averaged probabilities and RGB intensities might not be sufficient to show differences between the new pathology classes without the morphology context. I guess for my specific case of multi-class segmentation, the user needs to know for sure a certain class of object do exist in the image before providing the model the text prompt for it to avoid false positives.

**Comment by theodore-zhao** [comment]
Hi @WeisiX 

Thanks for the feedback! Just to make sure that the averaging is on the predicted area rather than on the whole image. Agree that subtle morphological difference can be challenging. We are also exploring improvements on that.
